src/main/java/mezz/jei/util/TagUtil.java

```diff
@@ -0,0 +1,35 @@
+package mezz.jei.util;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.ItemTags;
+import net.minecraft.tags.Tag;
+import net.minecraft.tags.TagCollection;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+public class TagUtil {
+	@Nullable
+	public static ResourceLocation getTagEquivalent(Collection<@Nullable ItemStack> itemStacks) {
```

Comment:
To make it general, I think this could be offloaded completely into something like 
`IIngredientHelper#getTagEquivalent(Collection<V> ingredients)`

The fluid handling is complicated so I've stuck with items only so far. There may be a good way to handle fluids in a custom way in its own `IIngredientHelper`.

Reply:
Added this as `IIngredientHelper#getTagEquivalent`